### PR TITLE
pluginupdate.py: fix bugs and add improvements; vimPlugins: sort properly

### PR DESCRIPTION
**Pull request by PerchunPak**
I fixed many hidden bugs and made some small improvements. The main reason this was separated from #336137 is to fix the sorting issue.

Before this commit, sorting for Vim plugins was broken and worked by what was fetched first. This is because the sorting was done by empty strings (the default value in CSV is not None but an empty string). This PR fixes it and also moves sorting from the user to the library (from `vim/plugins/update.py` to `pluginupdate.py`) to prevent such weird issues and duplication of code.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

CC @NixOS/neovim

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by GaetanLepage** [comment]
Thank you for this !
Could you please isolate the source code formatting in a separate commit please ? This would help with the review.

File: pkgs/applications/editors/kakoune/plugins/update.py [hunk 1]
```diff
@@ -2,47 +2,63 @@
...
-ROOT = Path(os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe())))) # type: ignore
+ROOT = Path(os.path.dirname(os.path.abspath(inspect.getfile(inspect.currentframe()))))  # type: ignore
 sys.path.insert(
     0, os.path.join(ROOT.parent.parent.parent.parent.parent, "maintainers", "scripts")
 )
 import pluginupdate
 
-GET_PLUGINS = f"""(with import <localpkgs> {{}};
+GET_PLUGINS = f"""(
+with import <localpkgs> {{ }};
```
Comment (PerchunPak) [inline pkgs/applications/editors/kakoune/plugins/update.py:25]: I also manually formatted this with `nixfmt-rfc-style`

File: pkgs/applications/editors/kakoune/plugins/update.py [hunk 2]
```diff
@@ -50,7 +50,7 @@
 lib.filterAttrs (n: v: v != null) checksums
 )"""
 
-HEADER = "# This file has been generated by ./pkgs/applications/editors/kakoune/plugins/update.py. Do not edit!"
+HEADER = "# This file has been @generated by ./pkgs/applications/editors/kakoune/plugins/update.py. Do not edit!"
```
Comment (GaetanLepage) [inline pkgs/applications/editors/kakoune/plugins/update.py:53]: why this `@` ?

**Review by GaetanLepage — COMMENTED** [review]
Thank you for splitting out the PR, it has been much easier to review it.
Also, this is a very useful work, good job :)

File: pkgs/applications/editors/kakoune/plugins/update.py (hunk 2 above)
Comment (PerchunPak) [reply pkgs/applications/editors/kakoune/plugins/update.py:53]: Just saw that the same does poetry

https://github.com/python-poetry/poetry/pull/2773

Comment (GaetanLepage) [reply pkgs/applications/editors/kakoune/plugins/update.py:53]: Ok, fine by me.

**Review by GaetanLepage — APPROVED** [review]
LGTM ! Good job :)

**Review by khaneliman — APPROVED** [review]
LGTM thanks for tackling this

**Comment by GaetanLepage** [comment]
## `nixpkgs-review` result

Generated using [`nixpkgs-review`](https://github.com/Mic92/nixpkgs-review).

Command: `nixpkgs-review pr 353786`

---
### `x86_64-linux`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>luarocks-packages-updater</li>
    <li>luarocks-packages-updater.dist</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

---
### `aarch64-linux`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>luarocks-packages-updater</li>
    <li>luarocks-packages-updater.dist</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

---
### `x86_64-darwin`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>luarocks-packages-updater</li>
    <li>luarocks-packages-updater.dist</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

---
### `aarch64-darwin`
<details>
  <summary>:white_check_mark: 3 packages built:</summary>
  <ul>
    <li>luarocks-packages-updater</li>
    <li>luarocks-packages-updater.dist</li>
    <li>vimPluginsUpdater</li>
  </ul>
</details>

**Comment by teto** [comment]
ran it locally. Seemed ok.

**Comment by teto** [comment]
thanks for the work. Now onto single plugin updates :p 
sry it took so much time but this could have been instantaneous if there were not 3 different changes bundled in the PR xD

**Pull request merged**